mkosi

```diff
@@ -2263,76 +2323,85 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
-
-        # Disable fallback images to save time. We override the preset files
-        # before installing any kernels to have the updated config available
-        # when the mkinitcpio pacman hook is run during installation.
-        presetdir = os.path.join(root, "etc/mkinitcpio.d")
-        for kernel in kernel_packages:
-            preset = os.path.join(presetdir, f"{kernel}.preset")
-            with open(preset, "w") as f:
-                f.write(
-                    dedent(
-                        f"""
-                        ALL_config="/etc/mkinitcpio.conf"
-                        ALL_kver="/boot/vmlinuz-{kernel}"
-                        PRESETS=('default')
-                        default_image="/boot/initramfs-{kernel}.img"
-                        """
-                    )
-                )
-
     # Install the user-specified packages and kernel
     packages = set(args.packages)
-    if not do_run_build_script and args.bootable:
-        packages |= kernel_packages
+    if not do_run_build_script and args.bootable and not any(["linux" in package for package in args.packages]):
+        # No user-specified kernel
+        packages.add("linux")
 
     if do_run_build_script:
         packages.update(args.build_packages)
 
-    def run_pacman(args: List[str], **kwargs: Any) -> CompletedProcess:
-        cmdline = [
-            "pacman",
-            "--noconfirm",
-            "--color", "never",
-            "--config", pacman_conf,
-        ]
+    if not do_run_build_script and args.bootable:
+        hooks_dir = os.path.join(root, "etc/pacman.d/hooks")
+        os.makedirs(hooks_dir, 0o755, exist_ok=True)
 
-        try:
-            return run(cmdline + args, **kwargs, check=True)
-        finally:
-            stop_gpg_agent()
+        # Disable depmod pacman hook as depmod is handled by kernel-install as well.
+        os.symlink("/dev/null", os.path.join(hooks_dir, "60-depmod.hook"))
 
-    # Remove already installed packages
-    c = run_pacman(['-Qq'], stdout=PIPE, universal_newlines=True)
-    packages.difference_update(c.stdout.split())
+        kernel_install_hook = os.path.join(hooks_dir, "kernel-install.hook")
+        with open(kernel_install_hook, 'w') as f:
+            f.write(
+                dedent(
+                    f"""
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Type = File
+                    Target = usr/lib/modules/*/vmlinuz
+
+                    [Action]
+                    Description = Adding kernel and initramfs images to /boot...
+                    When = PostTransaction
+                    Exec = /bin/bash -c 'while read f; do kernel-install add $(basename $(dirname $f)) $f; done'
+                    Depends = systemd
+                    Depends = dracut
+                    Depends = bash
+                    Depends = coreutils
```

Comment:
The depends on systemd, bash and coreutils is not strictly necessary, as they are dependencies of the base package and Arch only supports systems running base.

Reply:
Does it need to depend on dracut actually? It's a bit more complicated with the custom dracut hook we add but in general I don't think we need to add dracut here either since it installs its own kernel-install script.